### PR TITLE
migrate libbeat pipeline

### PR DESCRIPTION
**Pull request by sharbuz**
## What is the problem this PR solves?

Jenkins->Buildkite pipelines migration 

## How does this PR solve the problem?
1. Migrate the `libbeat` pipeline
https://buildkite.com/elastic/beats-libbeat/builds?branch=sharbuz%3Amigrate-libbeat-pipeline
2. Update the common conditions/rules for starting the pipelines (excluded the part from the main pipeline)

Examples of the builds:
Metricbeat: https://buildkite.com/elastic/beats-metricbeat/builds/1992
Libbeat: https://buildkite.com/elastic/beats-libbeat/builds/700

## Related issues

https://github.com/elastic/ingest-dev/issues/1693

### CODE REVIEW

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 1]
```diff
@@ -1,5 +1,76 @@
...
+    key: "extended-tests"
+    steps:
+      - label: ":mac: Arm64 Unit Tests"
+        key: "extended-arm64-unit-tests"
+        command: ".buildkite/scripts/unit_tests.sh ${PIPELINE_NAME}"
+        agents:
+          provider: "aws"
+          imagePrefix: "${IMAGE_UBUNTU_ARM_64}"
+          instanceType: "t4g.xlarge"
+        artifact_paths: "metricbeat/build/*.*"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:76]: is that the expected artifacts folder for libbeat?

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 2]
```diff
@@ -1,5 +1,76 @@
...
+
+      - label: ":negative_squared_cross_mark: Cross compile"
+        key: "mandatory-stress-test"
+        command: ".buildkite/scripts/stress_tests.sh ${PIPELINE_NAME}"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2-standard-16"
+        artifact_paths: "metricbeat/build/*.*"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:64]: is that the expected artifacts folder for libbeat?

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 3]
```diff
@@ -1,5 +1,76 @@
...
+
+      - label: ":python: Python Integration Tests"
+        key: "mandatory-python-int-test"
+        command: ".buildkite/scripts/py_int_tests.sh ${PIPELINE_NAME}"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2-standard-16"
+        artifact_paths: "metricbeat/build/*.*"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:46]: is that the expected artifacts folder for libbeat?

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 4]
```diff
@@ -1,5 +1,76 @@
...
+
+      - label: ":go: Go Intergration Tests"
+        key: "mandatory-int-test"
+        command: ".buildkite/scripts/go_int_tests.sh ${PIPELINE_NAME}"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2-standard-16"
+        artifact_paths: "metricbeat/build/*.*"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:37]: is that the expected artifacts folder for libbeat?

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 5]
```diff
@@ -1,5 +1,76 @@
...
+    key: "mandatory-tests"
+    steps:
+      - label: ":linux: Ubuntu Unit Tests"
+        key: "mandatory-linux-unit-test"
+        command: ".buildkite/scripts/unit_tests.sh ${PIPELINE_NAME}"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2-standard-16"
+        artifact_paths: "metricbeat/build/*.*"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:28]: is that the expected artifacts folder for libbeat?

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 6]
```diff
@@ -1,5 +1,76 @@
 # yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
+env:
+  SETUP_GVM_VERSION: "v0.5.1"
+  SETUP_MAGE_VERSION: "1.13.0"
+  SETUP_WIN_PYTHON_VERSION: "3.11.0"
+  IMAGE_UBUNTU_X86_64: "family/core-ubuntu-2204"
+  IMAGE_UBUNTU_ARM_64: "core-ubuntu-2004-aarch64"
+  IMAGE_WIN_2016: "family/core-windows-2016"
+  IMAGE_WIN_2019: "family/core-windows-2019"
+  IMAGE_WIN_2022: "family/core-windows-2022"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:10]: I cannot find where these env variables are used

File: .buildkite/libbeat/pipeline.libbeat.yml [hunk 7]
```diff
@@ -1,5 +1,76 @@
...
+  IMAGE_MACOS_X86_64: "generic-13-ventura-x64"
```
Comment (v1v) [inline .buildkite/libbeat/pipeline.libbeat.yml:11]: ditto

Comment (sharbuz) [reply .buildkite/libbeat/pipeline.libbeat.yml:11]: @v1v, It's just draft with the first view :) could you focus your efforts on the https://github.com/elastic/beats/pull/37592 PR? It's really important for now. Thanks!

File: .buildkite/scripts/crosscompile.sh [hunk 8]
```diff
@@ -7,4 +7,4 @@ set -euo pipefail
 beats_subfilder=$1
```
Comment (v1v) [inline .buildkite/scripts/crosscompile.sh:7]: typo `subfolder`

File: .buildkite/scripts/unit_tests.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+set -euo pipefail
+
+beats_subfolder=$1
+
+sudo chmod -R go-w ${beats_subfolder}/
+
+echo "--- Run Unit Tests"
+pushd "${beats_subfolder}" > /dev/null
+
+umask 0022
```
Comment (dliappis) [inline .buildkite/scripts/unit_tests.sh:14]: can we add a small `#TODO remove when custom vm images are used` as we have fixed this e.g. in the Logstash base images https://docs.google.com/document/d/1IDM9FEyrlrKmjZge3iBa-maaOFeK_ONjxLBMEXkwLI0/edit#bookmark=id.x7640ku6nmd3

(I assume this script will run on Linux, I am not sure if we need to also cover for macOS, in which case, custom images won't help, as we can't have custom images yet for the macos provider)

File: .buildkite/scripts/generate_metricbeat_pipeline.sh [hunk 10]
```diff
@@ -70,28 +67,8 @@ if are_conditions_met_mandatory_tests; then
...
-
-YAML
-fi
-
-if are_conditions_met_extended_windows_tests; then
-  cat >> $pipelineName <<- YAML
+# echo "Add the extended windows tests into the pipeline"
+# TODO: ADD conditions from the main pipeline
 
   - group: "Extended Windowds Tests"
```
[the remark this answers is not in the document]
Comment (sharbuz) [reply .buildkite/scripts/generate_metricbeat_pipeline.sh]: fixed

**Review by dliappis — APPROVED** [review]
LGTM provided CI is green

File: .buildkite/scripts/generate_libbeat_pipeline.sh [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      - label: ":linux: Ubuntu Unit Tests"
+        key: "mandatory-linux-unit-test"
+        command: ".buildkite/scripts/unit_tests.sh"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2-standard-16"
+        artifact_paths: "${BEATS_PROJECT_NAME}/build/*.xml"
+
+      - label: ":go: Go Intergration Tests"
```
[the remark this answers is not in the document]
Comment (sharbuz) [reply .buildkite/scripts/generate_libbeat_pipeline.sh]: fixed

File: .buildkite/scripts/generate_libbeat_pipeline.sh [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  - group: "Mandatory Tests"
+    key: "mandatory-tests"
+    steps:
+      - label: ":linux: Ubuntu Unit Tests"
+        key: "mandatory-linux-unit-test"
+        command: ".buildkite/scripts/unit_tests.sh"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2-standard-16"
```
[the remark this answers is not in the document]
Comment (sharbuz) [reply .buildkite/scripts/generate_libbeat_pipeline.sh]: changed:
https://elastic.slack.com/archives/C0522G6FBNE/p1707146144097869?thread_ts=1707140315.980879&cid=C0522G6FBNE

File: catalog-info.yaml [hunk 13]
```diff
@@ -322,11 +321,11 @@ spec:
           build.pull_request.id == null || (build.creator.name == 'elasticmachine' && build.pull_request.id != null)
       repository: elastic/beats
       cancel_intermediate_builds: true
-      cancel_intermediate_builds_branch_filter: "!main !7.17 !8.*"
+      cancel_intermediate_builds_branch_filter: "!main !7.* !8.*"
       skip_intermediate_builds: true
-      skip_intermediate_builds_branch_filter: "!main !7.17 !8.*"
-      # env:
-      #   ELASTIC_PR_COMMENTS_ENABLED: "true" TODO: uncomment when pipeline is ready
+      skip_intermediate_builds_branch_filter: "!main !7.* !8.*"
```
Comment (v1v) [inline catalog-info.yaml:326]: no need for supporting any 7.x but 7.17 
```suggestion
      skip_intermediate_builds_branch_filter: "!main !7.17 !8.*"
```

File: catalog-info.yaml [hunk 14]
```diff
@@ -310,9 +309,9 @@ spec:
       name: beats-libbeat
       description: "Beats libbeat pipeline"
     spec:
-#      branch_configuration: "main 7.17 8.* v7.17 v8.*" TODO: temporarily commented to build PRs from forks
+      branch_configuration: "main 7.* 8.* v7.* v8.*"
```
Comment (v1v) [inline catalog-info.yaml:312]: I'm not sure if tags should trigger any build, they are in fact managed by the Unified Release process and hence tag from commits that have been already tested it out.

```suggestion
      branch_configuration: "main 7.17 8.*
```

File: libbeat/scripts/Makefile [hunk 15]
```diff
@@ -46,7 +46,7 @@ export PATH := ./bin:$(PATH)
 GOFILES = $(shell find . -type f -name '*.go' 2>/dev/null)
 GOFILES_NOVENDOR = $(shell find . -type f -name '*.go' -not -path "*/vendor/*" 2>/dev/null)
 GOFILES_ALL = $(GOFILES) $(shell find $(ES_BEATS) -type f -name '*.go' 2>/dev/null)
-GOPACKAGES_STRESSTESTS=$(shell find . -name '*.go' 2>/dev/null | xargs grep -l '\+build.*stresstest' | xargs -n1 dirname | uniq)
+GOPACKAGES_STRESSTESTS=$(shell find . -type d \( -name "stress" \) 2>/dev/null)
```
Comment (v1v) [inline libbeat/scripts/Makefile:49]: What's the reason for this change?

File: libbeat/scripts/Makefile [hunk 16]
```diff
@@ -87,8 +87,9 @@ SYSTEM_TESTS?=false ## @testing if true, "make test" and "make testsuite" run un
 STRESS_TESTS?=false ## @testing if true, "make test" and "make testsuite" run also run the stress tests
 STRESS_TEST_OPTIONS?=-timeout=20m -race -v
 GOX_OS?=linux darwin windows freebsd netbsd openbsd ## @Building List of all OS to be supported by "make crosscompile".
-GOX_OSARCH?=!darwin/arm !darwin/arm64 ## @building Space separated list of GOOS/GOARCH pairs to build by "make crosscompile".
+GOX_OSARCH?=!darwin/arm !darwin/386 !linux/386 !windows/386 !freebsd/386 !netbsd/386 !openbsd/386 !linux/ppc64 ## @building Space separated list of GOOS/GOARCH pairs to build by "make crosscompile".
```
Comment (v1v) [inline libbeat/scripts/Makefile:90]: What's the reason for this change?

File: libbeat/scripts/Makefile [hunk 17]
```diff
@@ -87,8 +87,9 @@ SYSTEM_TESTS?=false ## @testing if true, "make test" and "make testsuite" run un
...
 GOX_FLAGS?= ## @building Additional flags to append to the gox command used by "make crosscompile".
+GOTEST_OUTPUT_OPTIONS?=
```
Comment (v1v) [inline libbeat/scripts/Makefile:92]: What's the reason for this new flag? A comment that explains what it does can help in the future.

File: .buildkite/scripts/generate_metricbeat_pipeline.sh [hunk 18]
```diff
@@ -33,7 +29,7 @@ if are_conditions_met_mandatory_tests; then
         agents:
           provider: "gcp"
           image: "${IMAGE_UBUNTU_X86_64}"
-          machineType: "c2-standard-16"
+          machineType: "c2d-highcpu-16"
```
Comment (v1v) [inline .buildkite/scripts/generate_metricbeat_pipeline.sh:32]: Since this value is hardcoded in a few places, maybe a variable could help.

File: .buildkite/scripts/generate_libbeat_pipeline.sh [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+          image: "${IMAGE_UBUNTU_X86_64}"
+        artifact_paths: "${BEATS_PROJECT_NAME}/build/*.xml"
+
+      - label: ":go: Go Integration Tests"
+        key: "mandatory-int-test"
+        command: ".buildkite/scripts/go_int_tests.sh"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
+          machineType: "c2d-highcpu-16"
```
Comment (v1v) [inline .buildkite/scripts/generate_libbeat_pipeline.sh:32]: Since this value is hardcoded in a few places, maybe a variable could help.

File: .buildkite/scripts/generate_libbeat_pipeline.sh [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+  - group: "Mandatory Tests"
+    key: "mandatory-tests"
+    steps:
+      - label: ":linux: Ubuntu Unit Tests"
+        key: "mandatory-linux-unit-test"
+        command: ".buildkite/scripts/unit_tests.sh"
+        agents:
+          provider: "gcp"
+          image: "${IMAGE_UBUNTU_X86_64}"
```
Comment (v1v) [inline .buildkite/scripts/generate_libbeat_pipeline.sh:23]: is `machineType` required in this particular case? Maybe4 it's a good idea to avoid relying on the provisioned but enforcing the required type of machine?

File: libbeat/scripts/Makefile (hunk 15 above)
Comment (sharbuz) [reply libbeat/scripts/Makefile:49]: Because the original line doesn't work at all. And, if I understood that line correctly, the main idea is to get paths to folders with stress tests. For the particular beat, I found only one place with stress tests, this place is covered by the new line...

File: libbeat/scripts/Makefile (hunk 17 above)
Comment (sharbuz) [reply libbeat/scripts/Makefile:92]: Because in the original pipeline stage are no artifacts at all. I use this option to see the result of tests as the artifacts in the build (define artifact path and then upload that to BK)

File: .buildkite/scripts/generate_metricbeat_pipeline.sh (hunk 18 above)
Comment (sharbuz) [reply .buildkite/scripts/generate_metricbeat_pipeline.sh:32]: I could be changed again https://elastic.slack.com/archives/C0522G6FBNE/p1707140315980879

File: .buildkite/scripts/generate_libbeat_pipeline.sh (hunk 20 above)
Comment (sharbuz) [reply .buildkite/scripts/generate_libbeat_pipeline.sh:23]: @v1v, please take a look at the test results https://elastic.slack.com/archives/C0522G6FBNE/p1707152502952139?thread_ts=1707140315.980879&cid=C0522G6FBNE

File: libbeat/scripts/Makefile (hunk 16 above)
Comment (sharbuz) [reply libbeat/scripts/Makefile:90]: because the latest version of `gox `generates too many different variants of OS/ARCH, and some of them are unsupported by GO.

Comment (v1v) [reply libbeat/scripts/Makefile:90]: Gotcha, I wonder whether this should be done in a different PR to avoid mixing concerns

File: libbeat/scripts/Makefile (hunk 15 above)
Comment (v1v) [reply libbeat/scripts/Makefile:49]: https://github.com/elastic/beats/pull/21719 was the original change about 4 years ago.

As far as I see it's defined in [here](https://github.com/elastic/beats/blob/12b73b7543fe0bd805be25b08a3b2dec3cb47ae6/libbeat/Jenkinsfile.yml#L39-L41) for Jenkins and it [ran](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/main/2287/pipeline/2388) without running:

```
24-02-05T16:13:59.159Z] + make STRESS_TEST_OPTIONS=-timeout=20m -race -v -parallel 1 -C libbeat stress-tests
[2024-02-05T16:13:59.159Z] make: Entering directory '/var/lib/jenkins/workspace/main-2287-c122c791-d81b-466c-8ad7-4cbbc004111d/src/github.com/elastic/beats/libbeat'
[2024-02-05T16:13:59.587Z] dirname: missing operand
[2024-02-05T16:13:59.588Z] Try 'dirname --help' for more information.
[2024-02-05T16:13:59.746Z] dirname: missing operand
[2024-02-05T16:13:59.746Z] Try 'dirname --help' for more information.
[2024-02-05T16:13:59.746Z] if [ -n '' ]; then \
[2024-02-05T16:13:59.746Z] 	go test --tags=stresstest -timeout=20m -race -v -parallel 1 ; \
[2024-02-05T16:13:59.746Z] fi
```

I wonder whether this should be fixed in a different PR to avoid mixing concerns

File: libbeat/scripts/Makefile (hunk 17 above)
Comment (v1v) [reply libbeat/scripts/Makefile:92]: Can you add a comment explaining that as similarly done in L91 or others?

File: catalog-info.yaml (hunk 13 above)
Comment (v1v) [reply catalog-info.yaml:326]: That won't be the case. `7.17` is the last branch for `7.x` - hence no `7.18` branch will be created.  https://www.elastic.co/support_policy explains what's the support.

File: libbeat/scripts/Makefile (hunk 17 above)
Comment (sharbuz) [reply libbeat/scripts/Makefile]: +

File: .buildkite/scripts/generate_metricbeat_pipeline.sh (hunk 18 above)
Comment (sharbuz) [reply .buildkite/scripts/generate_metricbeat_pipeline.sh]: done

File: .buildkite/scripts/generate_libbeat_pipeline.sh (hunk 19 above)
Comment (sharbuz) [reply .buildkite/scripts/generate_libbeat_pipeline.sh]: done

File: .buildkite/scripts/generate_libbeat_pipeline.sh (hunk 20 above)
Comment (sharbuz) [reply .buildkite/scripts/generate_libbeat_pipeline.sh:23]: done

File: libbeat/scripts/Makefile (hunk 16 above)
Comment (sharbuz) [reply libbeat/scripts/Makefile]: done, https://github.com/elastic/beats/pull/37870

File: libbeat/scripts/Makefile (hunk 15 above)
Comment (sharbuz) [reply libbeat/scripts/Makefile:49]: done, https://github.com/elastic/beats/pull/37873

**Review by v1v — APPROVED** [review]

**Review by jlind23 — APPROVED** [review]

**Comment by jlind23** [comment]
Force merging this as the failure is unrelated. @sharbuz will create a follow up issue for @axw and @bturquet to investigate.
